### DISCUSSION TITLE
Passed value cannot be an array - Fixed the issue, but raised another one related to Bard fields

### DISCUSSION THREAD

**Discussion by ItsJustPW** [statamic/cms#7414] [General]
I've been following Jacks course on [Laracasts](https://laracasts.com/series/learn-statamic-with-jack) and everything was going fine with the exception of the Bard field. I couldn't get it to render anything except plain text. I then realised I had broke something. As someone still trying to find their feet when it comes to developing... I checked the logs and seen this: 

`[2023-01-26 20:11:27] local.ERROR: Passed value cannot be an array {"userId":"7fca6675-5318-423a-9001-2b009693218d","exception":"[object] (Statamic\\Modifiers\\ModifierException(code: 0): Passed value cannot be an array at /Users/paul/Projects/ttg/vendor/statamic/cms/src/Modifiers/Modify.php:170)
[stacktrace]
#0 /Users/paul/Projects/ttg/vendor/statamic/cms/src/Modifiers/Modify.php(133): Statamic\\Modifiers\\Modify->modify('safe_truncate', Array)
#1 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/ModifierManager.php(119): Statamic\\Modifiers\\Modify->__call('safe_truncate', Array)
#2 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/Sandbox/Environment.php(1257): Statamic\\View\\Antlers\\Language\\Runtime\\ModifierManager::evaluate(Object(Statamic\\Fields\\Value), Object(Statamic\\View\\Antlers\\Language\\Runtime\\Sandbox\\Environment), Object(Statamic\\View\\Antlers\\Language\\Nodes\\Modifiers\\ModifierChainNode), Array)
#3 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/Sandbox/Environment.php(1278): Statamic\\View\\Antlers\\Language\\Runtime\\Sandbox\\Environment->applyModifiers(Object(Statamic\\Fields\\Value), Object(Statamic\\View\\Antlers\\Language\\Nodes\\Modifiers\\ModifierChainNode))
#4 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/Sandbox/Environment.php(1515): Statamic\\View\\Antlers\\Language\\Runtime\\Sandbox\\Environment->adjustValue(Object(Statamic\\Fields\\Value), Object(Statamic\\View\\Antlers\\Language\\Nodes\\VariableNode))
#5 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/Sandbox/Environment.php(923): Statamic\\View\\Antlers\\Language\\Runtime\\Sandbox\\Environment->getValue(Object(Statamic\\View\\Antlers\\Language\\Nodes\\VariableNode))
#6 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/Sandbox/Environment.php(296): Statamic\\View\\Antlers\\Language\\Runtime\\Sandbox\\Environment->process(Array)
#7 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/Sandbox/Environment.php(321): Statamic\\View\\Antlers\\Language\\Runtime\\Sandbox\\Environment->processStatements(Array)
#8 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/NodeProcessor.php(1727): Statamic\\View\\Antlers\\Language\\Runtime\\Sandbox\\Environment->evaluate(Array)
#9 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/NodeProcessor.php(712): Statamic\\View\\Antlers\\Language\\Runtime\\NodeProcessor->reduce(Array)
#10 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/RuntimeParser.php(382): Statamic\\View\\Antlers\\Language\\Runtime\\NodeProcessor->render(Array)
#11 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Language/Runtime/RuntimeParser.php(760): Statamic\\View\\Antlers\\Language\\Runtime\\RuntimeParser->renderText('<!-- article --...', Array)
#12 /Users/paul/Projects/ttg/vendor/statamic/cms/src/View/Antlers/Engine.php(97): Statamic\\View\\Antlers\\Language\\Runtime\\RuntimeParser->parseView('/Users/paul/Pro...', '<!-- article --...', Array)
#13 /Users/paul/Projects/ttg/vendor/laravel/framework/src/Illuminate/View/View.php(195): Statamic\\View\\Antlers\\Engine->get('/Users/paul/Pro...', Array)
#14 /Users/paul/Projects/ttg/vendor/laravel/framework/src/Illuminate/View/View.php(178): Illuminate\\View\\View->getContents()
#15 /Users/paul/Projects/ttg/vendor/laravel/framework/src/Illuminate/View/View.php(147): Illuminate\\View\\View->renderContents()
#16 /Users/paul/Projects/ttg/vendor/statamic/cms/src/Tags/Partial.php(29): Illuminate\\View\\View->render()
#17 /Users/paul/Projects/ttg/vendor/statamic/cms/src/Tags/Partial.php(13): Statamic\\Tags\\Partial->render('news/card')
#18 /Users/paul/Projects/ttg/vendor/statamic/cms/src/Tags/Tags.php(138): Statamic\\Tags\\Partial->wildcard('news/card')
#19 [internal function]: Statamic\\Tags\\Tags->__call('news/card', Array)`

It had my head battered as everything was working fine.  While it may be obvious to experienced developers... it took me a good hour to work out it was related to using truncate in a field which I had now turned into a bard field with options.  So, as a quick workaround I changed {{ news_text }} to be {{ title }} and it worked. Obviousy it looks stupid... but it was just to get it working until I could ask for help. So I'm not out of the woods just yet.

The Bard field seems to be my Kryptonite and it would be great if someone is able to explain/help with the following: 

1) Why was the news item only very limited formatting of text, even without preserving breaks:
Link to what it looks like in browser: [https://ibb.co/p1RyD7z](https://ibb.co/p1RyD7z)
Here's where I'm trying to display the field in my antlers file:
`{{ news_content}}
    {{ if type == "text" }}
        {{ text }}
    {{ elseif type == "quote" }}
        <div class="quote-block">
            &ldquo;{{ quote }}&rdquo;
            <div class="italic mt-4">&dash; {{ author }}</div>
        </div>
    {{  elseif type == "related_articles" }}
        {{ related_entries }}
            <a class="" href="{{ url }}">{{ title }}</a>
        {{ /related_entries }}
    {{ /if }}
{{ /news_content }}`

2) Does safe_truncate not work with a bard field - or am I using it wrong? 
3) Can anyone see anything I'm doing wrong which is causing the Bard field to not work as expected? 

I have read the Statamic docs, but I thought I did everything it was asking - plus following Jack in the videos, the only thing I can think of is I missed something. If anyone is able to offer help/advise/solutions that would be great thank you.

**Comment by jasonvarga** [comment]
I ran into that cryptic error recently, and [fixed it in 3.3.66](https://github.com/statamic/cms/pull/7409). That would have at least pointed to the truncate modifier faster.

> Why was the news item only very limited formatting of text, even without preserving breaks

Your screenshot looks like everything was being output correctly - you can tell by the italic and bold text mixed in, but you need some css to style it.

> Does safe_truncate not work with a bard field - or am I using it wrong?

It only works on plain text. When you have bard configured with sets, it'll be an array, which was the error you were getting. You could have used the `raw` and `bard_text` modifiers in your case. https://statamic.dev/modifiers/bard_text

`{{ bard_field | raw | bard_text | safe_truncate }}`

**Reply by ItsJustPW** [reply]
**Thank you** for taking the time to explain all of that and it has resolved my origional issue.

**Reply by ItsJustPW** [reply]
Thank you @jasonvarga I took your advise about styling and came up with some code to style it nice.  The output of which can be seen in [this screenshot](https://ibb.co/CB7RJbk). 
`                            <figure class="max-w-screen-md mx-auto text-center">
                                <svg aria-hidden="true" class="w-12 h-12 mx-auto mb-3 text-gray-400 dark:text-gray-600" viewBox="0 0 24 27" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M14.017 18L14.017 10.609C14.017 4.905 17.748 1.039 23 0L23.995 2.151C21.563 3.068 20 5.789 20 8H24V18H14.017ZM0 18V10.609C0 4.905 3.748 1.038 9 0L9.996 2.151C7.563 3.068 6 5.789 6 8H9.983L9.983 18L0 18Z" fill="currentColor"/></svg>
                                <blockquote>
                                    <p class="text-2xl italic font-medium text-white">This is a sample piece of text used for the quote</p>
                                </blockquote>
                                <figcaption class="flex items-center justify-center mt-6 space-x-3">
                                    <div class="flex items-center divide-x-2 divide-gray-500 dark:divide-gray-700">
                                        <cite class="pr-3 font-medium text-white">Micheal Gough</cite>
                                        <cite class="pl-3 text-sm font-light text-gold-main">CEO at Google</cite>
                                    </div>
                                </figcaption>
                            </figure>`

When I use `{{ quote }}` to replace the text in my template, it looks like [this screenshot.](https://ibb.co/9wYW9nh)

The text is still there, but a dark colour and it's also adding an additional " - It seems the text is ingoring the style or overwriting it? Is it possible to explain whats going on please?

**Reply by jasonvarga** [reply]
How are you putting that `<figure>` into bard?

**Reply by ItsJustPW** [reply]
I'm not... which leads me to believe I should be?

I'm OK building things in HTML but when it comes to hooking then up with a Bard field... I'm obviously fundamentally missing something.

Could you let me know what actions I should take and should I take these actions based on every element?

Really sorry to be wasting your time.

**Reply by jasonvarga** [reply]
For the "text" parts of Bard, you can't have custom html (without doing complicated PHP stuff). It's just going to output semantic html which you can style with css.

If you want custom html for quotes, what you should do is make a "set" with appropriate fields in it (the quote, author, company, etc). Then you can hook into it in your templates.

Also, remove the quote button from the bard config, so that your client is forced to use your quote set instead of Bard's standard quote.

```
{{ bard_field }}
  {{ if type == "text" }}
    {{ text }}
  {{ elseif type == "quote" }}
    put all that custom tailwind quote html in here, and sprinkle the fields within.
    {{ quote_text }}, {{ author }}, {{ company }}, etc.
  {{ /if }}
{{ /bard_field }} 
```

**Reply by ItsJustPW** [reply]
Jason... I can't thank you enough mate! I think it's finally clicked. I'll crack on with it!

Can I get you a coffee?

**Reply by jasonvarga** [reply]
Happy to help. 😄
